src/metabase/api/dashboard.clj

```diff
@@ -580,37 +646,65 @@
    [:series             {:optional true} [:maybe [:sequential map?]]]])
 
+(def ^:private UpdatedDashboardTab
+  [:map
+   ;; id can be negative, it indicates a new card and BE should create them
+   [:id   ms/Int]
+   [:name ms/NonBlankString]])
+
+(defn- update-dashboard-with-tabs!
+  [dashboard new-tabs new-cards]
+  (let [{:keys [temp->real-tab-ids
+                deleted-tab-ids]}  (do-update-tabs! dashboard (:ordered_tabs dashboard) new-tabs)
+        current-cards              (if (seq deleted-tab-ids)
+                                       (remove (fn [card]
+                                                 (contains? deleted-tab-ids (:dashboard_tab_id card)))
+                                               (:ordered_cards dashboard))
+                                       (:ordered_cards dashboard))
+        new-cards                  (cond->> new-cards
+                                     ;; fixup the temporary tab ids with the real ones
+                                     (seq temp->real-tab-ids)
+                                     (map (fn [card]
+                                            (if-let [real-tab-id (get temp->real-tab-ids (:dashboard_tab_id card))]
+                                              (assoc card :dashboard_tab_id real-tab-id)
+                                              card)))
+                                     ;; if there are deleted tabs, we need to remove the dashcards from those tabs
+                                     (seq deleted-tab-ids)
+                                     (remove #(contains? deleted-tab-ids (:dashboard_tab_id %))))]
+    (do-update-dashcards! dashboard current-cards new-cards)))
+
 (api/defendpoint PUT "/:id/cards"
-  "Update `Cards` on a Dashboard. Request body should have the form:
-
-    {:cards [{:id                 ... ; DashboardCard ID
-              :size_x             ...
-              :size_y             ...
-              :row                ...
-              :col                ...
-              :parameter_mappings ...
-              :series             [{:id 123
-                                    ...}]}
-             ...]}"
-  [id :as {{:keys [cards]} :body}]
+  "Update `Cards` and `Tabs` on a Dashboard. Request body should have the form:
+
+    {:cards        [{:id                 ... ; DashboardCard ID
+                     :size_x             ...
+                     :size_y             ...
+                     :row                ...
+                     :col                ...
+                     :parameter_mappings ...
+                     :series             [{:id 123
+                                           ...}]}
+                     ...]
+     :ordered_tabs [{:id       ... ; DashboardTab ID
+                     :name     ...}]}"
+  [id :as {{:keys [cards ordered_tabs]} :body}]
   {id    ms/PositiveInt
-   cards [:sequential UpdatedDashboardCard]}
+   cards [:maybe (ms/maps-with-unique-key [:sequential UpdatedDashboardCard] :id)]
```

Comment:
Why are `cards` now nilable?

Reply:
I was thinking about a dashboard without cards, but that seems too loose. I'll make it required.

Reply:
updated in 2f31939d9d